### PR TITLE
Support Objects in octet-stream codec #1099

### PR DESCRIPTION
**Pull request by derwehr**
See #1099

As a first step, I added `ex:bitLength` and `ex:bitOffset` to all methods of the codec and extended the tests accordingly.

I'll add support for the `object` schema type next. But because I already changed a lot of code, there might be some material for discussion.

### CODE REVIEW

File: packages/core/src/codecs/octetstream-codec.ts [hunk 1]
```diff
@@ -115,23 +136,24 @@ export default class OctetstreamCodec implements ContentCodec {
...
         bytes: Buffer,
-        options: { dataLength: number; bigendian: boolean; signed: boolean }
+        options: { dataLength: number; bigEndian: boolean; signed: boolean }
     ): number {
-        const { dataLength, bigendian, signed } = options;
+        const { dataLength, bigEndian, signed } = options;
+
         switch (dataLength) {
-            case 1:
+            case 8:
```
Comment (danielpeintner) [inline packages/core/src/codecs/octetstream-codec.ts:144]: Out of curiosity. What is the big win using bit dataLength vs. byte dataLength?

Comment (derwehr) [reply packages/core/src/codecs/octetstream-codec.ts:142]: In the `integerToValue` method, the only point is to be consistent with using bit length throughout the codec where possible.
Further up at line 105, when reading only the bits specified at the bit dataLength, I am ensuring that the bit offset is a multiple of 8, so I could use the byte dataLength here, too.

**Comment by derwehr** [comment]
I add a few more test cases to increase coverage
Would you like me to move the new features to another codec instead of extending `octet-steam`, or is there anything else I can change to possibly get this merged?

**Review by danielpeintner — APPROVED** [review]
I think it is definitely an improvement since we now support objects  (once a schema is provided for it)

Note: I only did a quick review but it looks good to me and the additional tests are useful also!

Let's wait till @relu91 or @JKRhb  have time to take a look.

BTW, Sorry for not reacting for such a long time

**Comment by derwehr** [comment]
Thanks for the review.
By the way, one notable change I made is that I moved `byteSeq`, `signed` and `length` from the URI parameters into the data schema.
I'm unsure if you'd want that, but it seemed more natural. On second thought, that's a deprecation of the parameters, and maybe I should re-add them along with a check making sure they don't differ from the parameters provided by the data schema.
Please let me know which way you'd prefer.

**Comment by danielpeintner** [comment]
> By the way, one notable change I made is that I moved `byteSeq`, `signed` and `length` from the URI parameters into the data schema.
> I'm unsure if you'd want that, but it seemed more natural. On second thought, that's a deprecation of the parameters, and maybe I should re-add them along with a check making sure they don't differ from the parameters provided by the data schema.

@relu91 @egekorkan @JKRhb do you have an opinion on that aspect?

I think the advantage of having these parameters as URI means one can tweak them at runtime while in the schema case they are fixed. I would even argue that they can be at both places and the URI parameters takes precedence if available.. but I guess this makes it more complex.

**Comment by egekorkan** [comment]
> @relu91 @egekorkan @JKRhb do you have an opinion on that aspect?

Sadly none of the direction seems to be direction that the standards are developing towards. We will probably have protocol specific keywords for that but I can see why you would want to have these protocol independent since data encoding/coding happens in codec in node-wot, which is not aware of the protocol. I would definitely wait for @relu91 before merging.

**Comment by relu91** [comment]
Hi @derwehr! 👋
First of all, thank you very much for your efforts!! I like the idea of improving the octet-stream codec with the ability to encode/decode objects. Sorry for all the friction, but this is a delicate topic because it is a moving target. The problem is that the Thing Description and Protocol binding task forces (part of the WoT WG) are discussing, **these days**, how to improve the mapping between IoT protocol payloads to application data. In practice, in the short future, we might need to change how payloads are processed in node-wot. 

Having said that we discussed it internally and decided that for now, it is good to merge this PR and leave the refactoring for later. Before merging though I might ask you one final job. Let me clarify one point. `byteSeq`, `signed`, and `length` were not URI parameters but rather **content-type** parameters. As the RFC specifies, content-types might have parameters, and third argument of the `bytesToValue` and `valueToBytes` functions is the content-type parameters parsed by the upper layer. See [here](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/core/src/content-serdes.ts#L140). Therefore, I would ask you if it is possible to revert the change you made and keep those in the content-type parameters. 

Now, I'll do a review of the rest of your implementation and let you know if there is something else to change before merging.

File: packages/core/src/codecs/octetstream-codec.ts [hunk 2]
```diff
@@ -140,98 +166,132 @@ export default class OctetstreamCodec implements ContentCodec {
...
         }
     }
 
+    private objectToValue(bytes: Buffer, schema?: DataSchema): DataSchemaValue {
+        if (schema?.type !== "object") {
+            throw new Error("Schema must be of type 'object'");
+        }
+
+        const result: { [key: string]: unknown } = {};
+        for (const propertyName in schema.properties) {
```
Comment (relu91) [inline packages/core/src/codecs/octetstream-codec.ts:207]: to make this work the list of properties should always have the same order. Are we sure that this is consistent across different node instances? maybe it is safer to order them in alphabetical order?

File: packages/core/src/codecs/octetstream-codec.ts [hunk 3]
```diff
@@ -140,98 +166,132 @@ export default class OctetstreamCodec implements ContentCodec {
...
+            const propertySchema = schema.properties[propertyName];
+            result[propertyName] = this.bytesToValue(bytes, propertySchema);
```
Comment (relu91) [inline packages/core/src/codecs/octetstream-codec.ts:209]: Now I think I understand why you moved the parameters to the schema. I still keep then as global parameters in the content-type and then apply them to all the subschema. It is very unlikely that the same platform will encode the properties in a different format. If it is BE it is BE for all the payloads.

**Review by relu91 — CHANGES_REQUESTED** [review]
I skim through the code but it is hard to visually remove the changes that you did for the parameters. I left a couple of comments, but I think I review the code once again when you revert back to the previous handling of content-type parameters. Thanks!

File: packages/core/src/codecs/octetstream-codec.ts (hunk 2 above)
Comment (derwehr) [reply packages/core/src/codecs/octetstream-codec.ts]: done :)

File: packages/core/src/codecs/octetstream-codec.ts [hunk 4]
```diff
@@ -59,34 +59,57 @@ export default class OctetstreamCodec implements ContentCodec {
         debug("OctetstreamCodec parsing", bytes);
         debug("Parameters", parameters);
 
-        const bigendian = !(parameters.byteSeq?.includes(Endianness.LITTLE_ENDIAN) === true); // default to big endian
+        console.debug(parameters);
```
Comment (danielpeintner) [inline packages/core/src/codecs/octetstream-codec.ts:62]: Please remove or user proper logging

File: packages/core/src/codecs/octetstream-codec.ts (hunk 2 above)
Comment (derwehr) [reply packages/core/src/codecs/octetstream-codec.ts]: I now updated this line to use `Object.getOwnPropertyNames`, which is ordered in Node version 6.4 and up, see https://node.green/#ES2015-misc-own-property-order

File: packages/core/src/codecs/octetstream-codec.ts (hunk 3 above)
Comment (relu91) [reply packages/core/src/codecs/octetstream-codec.ts]: Yes please keep everything in the content-type for now.

Comment (relu91) [reply packages/core/src/codecs/octetstream-codec.ts]: I'd use the length parameter defined in the content type as an array length. If in the schema we are trying to access outside its ranges (either by specifying a bitOffset or a bitLenght) then it is a parsing error. If we want to downgrade gracefully we can take whatever is possible and log a warning.

Comment (derwehr) [reply packages/core/src/codecs/octetstream-codec.ts]: Perfect, I think that's taken care of by my most recent commit (28b135c346b922a6a22c7c34c8fd9324dda353a4) then

**Review by relu91 — APPROVED** [review]
Thank you very, much for your efforts! We can merge it and try to play with it. As a follow-up work, we should provide a short explanation about the new terms introduced.

**Comment by danielpeintner** [comment]
Thank you @derwehr for your contribution 👍

I created https://github.com/eclipse-thingweb/node-wot/issues/1194 as follow-up.

**Pull request merged**